File: capture_working_final.py
#!/usr/bin/env python3
"""
Working Final Capture Script
Uses the discovered coroutine classes for unlock/lock operations
"""

import frida
import sys
import json
import time
import subprocess
from datetime import datetime

PACKAGE_NAME = "fr.mayndrive.app"
SCRIPT_FILE = "capture_WORKING_FINAL.js"
OUTPUT_FILE = "CAPTURED_WORKING_FINAL.txt"
OUTPUT_JSON = "CAPTURED_WORKING_FINAL.json"

captured_data = []

def append_record(record):
    captured_data.append(record)
    with open(OUTPUT_JSON, 'w', encoding='utf-8') as jf:
        json.dump(captured_data, jf, indent=2, ensure_ascii=False)

def write_block(title, lines):
    with open(OUTPUT_FILE, 'a', encoding='utf-8') as f:
        separator = '=' * 100
        f.write(f"\n{separator}\n")
        f.write(f"{title}\n")
        f.write(f"{separator}\n")
        for line in lines:
            f.write(f"{line}\n")
        f.write(f"{separator}\n")

def restart_frida_server():
    print("[!] Restarting frida-server on device...")
    commands = [
        ["adb", "shell", "su", "-c", "pkill frida-server"],
        ["adb", "shell", "su", "-c", "/data/local/tmp/frida-server &"]
    ]
    for cmd in commands:
        try:
            subprocess.run(cmd, stdout=subprocess.PIPE, stderr=subprocess.PIPE, check=False)
        except FileNotFoundError:
            print("[-] adb not found in PATH. Please restart frida-server manually.")
            return False
    time.sleep(2)
    print("[+] frida-server restart issued")
    return True

def on_message(message, data):
    if message['type'] != 'send':
        if message['type'] == 'error':
            print(f"[ERROR] {message.get('description', message)}")
        return
    
    payload = message['payload']
    timestamp = datetime.now().isoformat()
    
    if payload.get('type') == 'unlock_request':
        print(f"\n[UNLOCK] {timestamp}")
        print(f"  Token: {payload.get('authorization', 'N/A')[:60]}...")
        print(f"  Scooter ID: {payload.get('scooter_id', 'N/A')}")
        print(f"  Location: {payload.get('location', 'N/A')}")
        print(f"  Class: {payload.get('className', 'N/A')}")
        
        # Save to file
        lines = [
            f"Authorization: {payload.get('authorization', 'N/A')}",
            f"Scooter ID: {payload.get('scooter_id', 'N/A')}",
            f"Location: {payload.get('location', 'N/A')}",
            f"Class: {payload.get('className', 'N/A')}"
        ]
        write_block(f"[{timestamp}] UNLOCK REQUEST", lines)
        
        # Save to JSON
        append_record(payload)
        
    elif payload.get('type') == 'lock_request':
        print(f"\n[LOCK] {timestamp}")
        print(f"  Token: {payload.get('authorization', 'N/A')[:60]}...")
        print(f"  Pass ID: {payload.get('pass_id', 'N/A')}")
        print(f"  Class: {payload.get('className', 'N/A')}")
        
        # Save to file
        lines = [
            f"Authorization: {payload.get('authorization', 'N/A')}",
            f"Pass ID: {payload.get('pass_id', 'N/A')}",
            f"Class: {payload.get('className', 'N/A')}"
        ]
        write_block(f"[{timestamp}] LOCK REQUEST", lines)
        
        # Save to JSON
        append_record(payload)
        
    elif payload.get('type') == 'telemetry_data':
        print(f"\n[TELEMETRY] {timestamp}")
        print(f"  Data: {payload.get('data', 'N/A')}")
        
        # Save to file
        lines = [f"Data: {payload.get('data', 'N/A')}"]
        write_block(f"[{timestamp}] TELEMETRY DATA", lines)
        
        # Save to JSON
        append_record(payload)

def main():
    print("=" * 100)
    print("WORKING FINAL CAPTURE - Using Discovered Coroutine Classes")
    print("=" * 100)
    print("This script captures unlock/lock requests using the real coroutine classes")
    print("discovered through automatic analysis.")
    print("=" * 100)
    
    try:
        # Get device
        device = frida.get_usb_device()
        print(f"[+] Device: {device.name}")
        
        # Load script
        with open(SCRIPT_FILE, 'r', encoding='utf-8') as f:
            script_code = f.read()
        
        # Clear output files
        open(OUTPUT_FILE, 'w').close()
        open(OUTPUT_JSON, 'w').close()
        
        # Spawn app
        print(f"[*] Spawning {PACKAGE_NAME}...")
        try:
            pid = device.spawn([PACKAGE_NAME])
        except frida.ProcessNotFoundError:
            print(f"[ERROR] App {PACKAGE_NAME} not found. Make sure it's installed.")
            return
        except frida.ProcessNotRespondingError:
            print(f"[ERROR] App {PACKAGE_NAME} not responding. Trying to restart frida-server...")
            if restart_frida_server():
                try:
                    pid = device.spawn([PACKAGE_NAME])
                except Exception as e:
                    print(f"[ERROR] Still failed to spawn app: {e}")
                    return
            else:
                return
        
        session = device.attach(pid)
        
        # Create script
        script = session.create_script(script_code)
        script.on('message', on_message)
        script.load()
        
        # Resume app
        device.resume(pid)
        print(f"[+] App launched! PID: {pid}")
        
        print("\n" + "=" * 100)
        print("READY TO CAPTURE!")
        print("=" * 100)
        print("Instructions:")
        print("1. Use the spawned app instance (don't switch apps)")
        print("2. Login to your account")
        print("3. Find a scooter")
        print("4. Press UNLOCK - watch for [UNLOCK] messages")
        print("5. Press LOCK - watch for [LOCK] messages")
        print("6. Press Ctrl+C when done")
        print("=" * 100)
        
        # Keep running
        sys.stdin.read()
        
    except KeyboardInterrupt:
        print("\n[OK] Capture stopped")
    except Exception as e:
        print(f"\n[ERROR] {e}")
    finally:
        if 'session' in locals():
            session.detach()
        print(f"\nText log: {OUTPUT_FILE}")
        print(f"JSON export: {OUTPUT_JSON}")
        print(f"Items captured: {len(captured_data)}")

if __name__ == "__main__":
    main()

File: capture_WORKING_FINAL.js
/**
 * WORKING FINAL CAPTURE - Uses discovered coroutine classes
 * Based on automatic discovery results
 */
'use strict';

const LOG_LINE = '='.repeat(96);

function unwrapField(owner, name) {
    try {
        var field = owner[name];
        if (field === null || field === undefined) return null;
        if (typeof field === 'object' && field !== null && 'value' in field) {
            try { return field.value; } catch (e) { return field; }
        }
        return field;
    } catch (e) {
        return null;
    }
}

function toStringSafe(value) {
    if (value === null || value === undefined) {
        return null;
    }
    try {
        return value.toString();
    } catch (e) {
        return null;
    }
}

function logBlock(title, lines) {
    console.log('\n' + LOG_LINE);
    console.log(title);
    if (lines && lines.length) {
        lines.forEach(function(line) {
            console.log(line);
        });
    }
    console.log(LOG_LINE + '\n');
}

Java.perform(function() {
    console.log('[*] WORKING FINAL CAPTURE - Installing discovered hooks\n');
    
    // UNLOCK COROUTINES (with location)
    var unlockClasses = ['B4.i1', 'B4.d3'];
    
    unlockClasses.forEach(function(className) {
        try {
            var clazz = Java.use(className);
            var invokeSuspend = clazz.invokeSuspend.overload('java.lang.Object');
            invokeSuspend.implementation = function(arg) {
                var token = toStringSafe(unwrapField(this, 'Z'));
                var location = unwrapField(this, 'g0');
                var scooterId = unwrapField(this, 'f0');
                
                var lines = [];
                if (token) lines.push('  Authorization: ' + token.substring(0, 60) + '...');
                if (scooterId) lines.push('  Scooter ID: ' + scooterId);
                if (location) lines.push('  Location: ' + location);
                
                logBlock('[UNLOCK] ' + className + ' - Scooter Unlock Request', lines);
                
                // Send to Python
                send({
                    type: 'unlock_request',
                    authorization: token,
                    scooter_id: scooterId ? scooterId.toString() : null,
                    location: location ? location.toString() : null,
                    timestamp: new Date().toISOString(),
                    className: className
                });
                
                return invokeSuspend.call(this, arg);
            };
            console.log('[+] Hooked ' + className + ' (UNLOCK)');
        } catch (e) {
            console.log('[-] Failed to hook ' + className + ': ' + e);
        }
    });
    
    // LOCK COROUTINES (token only)
    var lockClasses = ['B4.P3', 'B4.x', 'B4.q2', 'B4.U4', 'B4.r1'];
    
    lockClasses.forEach(function(className) {
        try {
            var clazz = Java.use(className);
            var invokeSuspend = clazz.invokeSuspend.overload('java.lang.Object');
            invokeSuspend.implementation = function(arg) {
                var token = toStringSafe(unwrapField(this, 'Z'));
                var passId = unwrapField(this, 'k0') || unwrapField(this, 'g0') || unwrapField(this, 'f0');
                
                var lines = [];
                if (token) lines.push('  Authorization: ' + token.substring(0, 60) + '...');
                if (passId) lines.push('  Pass ID: ' + passId);
                
                logBlock('[LOCK] ' + className + ' - Scooter Lock Request', lines);
                
                // Send to Python
                send({
                    type: 'lock_request',
                    authorization: token,
                    pass_id: passId ? passId.toString() : null,
                    timestamp: new Date().toISOString(),
                    className: className
                });
                
                return invokeSuspend.call(this, arg);
            };
            console.log('[+] Hooked ' + className + ' (LOCK)');
        } catch (e) {
            console.log('[-] Failed to hook ' + className + ': ' + e);
        }
    });
    
    // UDP Cipher hooks for telemetry decryption
    try {
        var Cipher = Java.use('javax.crypto.Cipher');
        var doFinal = Cipher.doFinal.overload('[B');
        doFinal.implementation = function(input) {
            var result = doFinal.call(this, input);
            
            // Check if this looks like telemetry data
            if (input && input.length > 10) {
                try {
                    var inputStr = Java.use('java.lang.String').$new(input, 'UTF-8');
                    if (inputStr.indexOf('lat') !== -1 || inputStr.indexOf('lon') !== -1 || 
                        inputStr.indexOf('battery') !== -1 || inputStr.indexOf('speed') !== -1) {
                        
                        console.log('\n' + LOG_LINE);
                        console.log('[UDP PLAINTEXT] Decrypted telemetry data:');
                        console.log(inputStr);
                        console.log(LOG_LINE + '\n');
                        
                        send({
                            type: 'telemetry_data',
                            data: inputStr,
                            timestamp: new Date().toISOString()
                        });
                    }
                } catch (e) {
                    // Not text data, ignore
                }
            }
            
            return result;
        };
        console.log('[+] Hooked Cipher.doFinal for telemetry decryption');
    } catch (e) {
        console.log('[-] Failed to hook Cipher: ' + e);
    }
    
    // Native UDP hooks
    try {
        var sendto = Module.getExportByName(null, 'sendto');
        if (sendto) {
            Interceptor.attach(sendto, {
                onEnter: function(args) {
                    this.socket = args[0].toInt32();
                    this.buf = args[1];
                    this.len = args[2].toInt32();
                },
                onLeave: function(retval) {
                    if (this.len > 0 && this.len < 2000) {
                        var data = this.buf.readByteArray(this.len);
                        var hex = '';
                        for (var i = 0; i < Math.min(data.length, 64); i++) {
                            hex += ('0' + data[i].toString(16)).slice(-2) + ' ';
                        }
                        
                        console.log('\n' + LOG_LINE);
                        console.log('[UDP SEND] Socket: ' + this.socket + ' Length: ' + this.len + ' bytes');
                        console.log(hex);
                        console.log(LOG_LINE + '\n');
                    }
                }
            });
            console.log('[+] Hooked native sendto');
        }
    } catch (e) {
        console.log('[-] Failed to hook sendto: ' + e);
    }
    
    console.log('\n[*] ALL HOOKS INSTALLED!\n');
    console.log('[!] Ready to capture unlock/lock requests and telemetry data\n');
});
